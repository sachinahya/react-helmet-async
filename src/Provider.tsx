import React, { FC, ReactNode } from 'react';
import { HelmetStateClient } from './HelmetState';

export const Context = React.createContext<HelmetStateClient>({} as HelmetStateClient);

export interface HelmetProviderProps {
  state: HelmetStateClient;
  children?: ReactNode;
}

export const HelmetProvider: FC<HelmetProviderProps> = ({ state, children }) => {
  return <Context.Provider value={state}>{children}</Context.Provider>;
};
