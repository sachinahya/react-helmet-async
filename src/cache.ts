import { HelmetProps } from './state';

export interface HelmetCache {
  update(instance: unknown, props: HelmetProps): void;
  remove(instance: unknown): void;
}
