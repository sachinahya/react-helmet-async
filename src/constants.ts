import { LiteralUnion } from 'type-fest';

export const TAG_PROPERTIES = {
  CHARSET: 'charset',
  CSS_TEXT: 'cssText',
  HREF: 'href',
  HTTPEQUIV: 'http-equiv',
  INNER_HTML: 'innerHTML',
  ITEM_PROP: 'itemprop',
  NAME: 'name',
  PROPERTY: 'property',
  REL: 'rel',
  SRC: 'src',
} as const;

export const ATTRIBUTE_NAMES = {
  BODY: 'bodyAttributes',
  HTML: 'htmlAttributes',
  TITLE: 'titleAttributes',
} as const;

export const TAG_NAMES = {
  BASE: 'base',
  BODY: 'body',
  HEAD: 'head',
  HTML: 'html',
  LINK: 'link',
  META: 'meta',
  NOSCRIPT: 'noscript',
  SCRIPT: 'script',
  STYLE: 'style',
  TITLE: 'title',
  FRAGMENT: 'Symbol(react.fragment)',
} as const;

export const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['generator', 'robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
} as const;

export const VALID_TAG_NAMES = Object.values(TAG_NAMES);

export const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
} as const;

const swapKeysAndValues = <T extends Record<string | number | symbol, string | number | symbol>>(
  obj: T
): { [K in keyof T as T[K]]: K } => {
  return Object.fromEntries(Object.entries(obj).map(([key, value]) => [value, key])) as {
    [K in keyof T as T[K]]: K;
  };
};

export const HTML_TAG_MAP = swapKeysAndValues(REACT_TAG_MAP);

export const HELMET_ATTRIBUTE = 'data-rh';
