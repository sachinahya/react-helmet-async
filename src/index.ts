export { Helmet } from './Helmet';
export { type HelmetStateClient as HelmetState } from './HelmetState';
