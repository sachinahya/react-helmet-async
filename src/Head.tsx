import {
  Children,
  Component,
  ContextType,
  PropsWithChildren,
  ReactElement,
  ReactNode,
  isValidElement,
} from 'react';
import { isFragment } from 'react-is';
import fastCompare from 'react-fast-compare';
import invariant from 'tiny-invariant';
import { HeadContext } from './HeadContext';
import { NON_SELF_CLOSING_TAGS, TAG_NAMES, VALID_TAG_NAMES } from './constants';
import { HeadProps } from './state';

export interface HeadComponentProps {
  children?: ReactNode;
}

function flattenArrayTypeChildren(
  arrayTypeChildren: Record<string, unknown[]>,
  child: ReactElement<PropsWithChildren>
): void {
  if (typeof child.type !== 'string') {
    return;
  }

  const newChildProps = { ...child.props };

  if (newChildProps.children) {
    const isSelfClosing = !NON_SELF_CLOSING_TAGS.includes(child.type);

    if (isSelfClosing) {
      throw new Error(
        `<${child.type} /> elements are self-closing and can not contain children. Refer to our API for more information.`
      );
    }
  }

  (arrayTypeChildren[child.type] ??= []).push(newChildProps);
}

function mapObjectTypeChildren(
  mappedProps: HeadProps,
  child: ReactElement<PropsWithChildren>
): HeadProps {
  const { children: nestedChildren, ...newChildProps } = child.props;

  switch (child.type) {
    case TAG_NAMES.TITLE: {
      // eslint-disable-next-line no-nested-ternary
      mappedProps.title = Array.isArray(nestedChildren)
        ? nestedChildren
        : nestedChildren
        ? String(nestedChildren)
        : undefined;

      mappedProps.titleAttributes = newChildProps;

      break;
    }

    case TAG_NAMES.BODY:
      mappedProps.bodyAttributes = newChildProps;
      break;

    case TAG_NAMES.HTML:
      mappedProps.htmlAttributes = newChildProps;
      break;

    default: {
      // eslint-disable-next-line no-lone-blocks
      if (typeof child.type === 'string') {
        (mappedProps as Record<string, unknown>)[child.type] = newChildProps;
      }

      break;
    }
  }

  return mappedProps;
}

function mapChildrenToProps(componentProps: Readonly<HeadComponentProps>): HeadProps {
  const arrayTypeChildren: Record<string, unknown[]> = {};

  const { children, ...mappedProps } = componentProps;

  Children.forEach(children as ReactElement<PropsWithChildren>[], child => {
    if (!isValidElement(child)) {
      return;
    }

    // isFragment incorrectly narrows to ReactElement.
    // A child can still be a ReactElement if it's not a fragment.
    if (isFragment(child) as boolean) {
      Object.assign(
        mappedProps,
        mapChildrenToProps({
          ...mappedProps,
          children: child.props.children,
        })
      );
    } else {
      invariant(
        VALID_TAG_NAMES.some(name => child.type === name),
        typeof child.type === 'function'
          ? `You may be attempting to nest <Head> components within each other, which is not allowed. Refer to our API for more information.`
          : `Only elements types ${VALID_TAG_NAMES.join(
              ', '
            )} are allowed. Head does not support rendering <${
              child.type
            }> elements. Refer to our API for more information.`
      );

      invariant(
        !child.props.children ||
          typeof child.props.children === 'string' ||
          (Array.isArray(child.props.children) &&
            !child.props.children.some(nestedChild => typeof nestedChild !== 'string')),
        `Head expects a string as a child of <${child.type}>. Did you forget to wrap your children in braces? ( <${child.type}>{\`\`}</${child.type}> ) Refer to our API for more information.`
      );

      switch (child.type) {
        case TAG_NAMES.BASE:
        case TAG_NAMES.LINK:
        case TAG_NAMES.META:
        case TAG_NAMES.NOSCRIPT:
        case TAG_NAMES.SCRIPT:
        case TAG_NAMES.STYLE:
          flattenArrayTypeChildren(arrayTypeChildren, child);
          break;

        default:
          mapObjectTypeChildren(mappedProps, child);
          break;
      }
    }
  });

  for (const [arrayChildName, arrayChild] of Object.entries(arrayTypeChildren)) {
    (mappedProps as Record<string, unknown[]>)[arrayChildName] = arrayChild;
  }

  return mappedProps;
}

const assertContext = (context: ContextType<typeof HeadContext>) => {
  invariant(context, 'Missing HeadProvider');
  return context;
};

export class Head extends Component<HeadComponentProps> {
  static override contextType = HeadContext;

  // @ts-expect-error
  override context!: ContextType<typeof HeadContext>;

  rendered: boolean = false;

  override shouldComponentUpdate(nextProps: HeadComponentProps) {
    return !fastCompare(this.props, nextProps);
  }

  override componentDidUpdate() {
    const mappedProps = mapChildrenToProps(this.props);
    assertContext(this.context).update(this, mappedProps);
  }

  override componentWillUnmount() {
    assertContext(this.context).remove(this);
  }

  override render() {
    if (!this.rendered) {
      this.rendered = true;

      const mappedProps = mapChildrenToProps(this.props);
      assertContext(this.context).update(this, mappedProps);
    }

    return null;
  }
}
