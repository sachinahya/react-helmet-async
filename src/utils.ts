import { UnionToIntersection } from 'type-fest';
import { BodyProps, HelmetProps, HtmlProps } from './Helmet';
import { TAG_NAMES, TAG_PROPERTIES, ATTRIBUTE_NAMES } from './constants';

const HELMET_PROPS = {
  DEFAULT_TITLE: 'defaultTitle',
  DEFER: 'defer',
  ENCODE_SPECIAL_CHARACTERS: 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE: 'onChangeClientState',
  TITLE_TEMPLATE: 'titleTemplate',
  PRIORITIZE_SEO_TAGS: 'prioritizeSeoTags',
} as const;

const getInnermostProperty = <T extends keyof HelmetProps>(
  propsList: HelmetProps[],
  property: T
): HelmetProps[T] | undefined => {
  for (const props of [...propsList].reverse()) {
    if (props[property] != null) {
      return props[property];
    }
  }

  return undefined;
};

const getTitleFromPropsList = (propsList: HelmetProps[]): string | undefined => {
  const innermostTitle = getInnermostProperty(propsList, TAG_NAMES.TITLE);

  const innermostTemplate = getInnermostProperty(propsList, HELMET_PROPS.TITLE_TEMPLATE);

  if (innermostTemplate && innermostTitle) {
    // use function arg to avoid need to escape $ characters
    return innermostTemplate.replace(/%s/g, () => innermostTitle ?? '');
  }

  return innermostTitle || getInnermostProperty(propsList, HELMET_PROPS.DEFAULT_TITLE) || undefined;
};

const getOnChangeClientState = (propsList: HelmetProps[]): HelmetProps['onChangeClientState'] => {
  return getInnermostProperty(propsList, HELMET_PROPS.ON_CHANGE_CLIENT_STATE);
};

const getAttributesFromPropsList = <
  T extends 'bodyAttributes' | 'htmlAttributes' | 'titleAttributes'
>(
  tagType: T,
  propsList: HelmetProps[]
): NonNullable<HelmetProps[T]> => {
  const mergedAttributes: NonNullable<HelmetProps[T]> = {};

  for (const props of propsList) {
    if (props[tagType] != null) {
      Object.assign(mergedAttributes, props[tagType]);
    }
  }

  return mergedAttributes;
};

const getBaseTagFromPropsList = (
  propsList: HelmetProps[]
): React.JSX.IntrinsicElements['base'][] => {
  for (const props of [...propsList].reverse()) {
    if (props.base?.href) {
      return [props.base];
    }
  }

  return [];
};

// eslint-disable-next-line no-console
const warn = msg => console && typeof console.warn === 'function' && console.warn(msg);

const getTagsFromPropsList = <T extends 'link' | 'meta' | 'noscript' | 'script' | 'style'>(
  tagName: T,
  primaryAttributes: string[],
  propsList: HelmetProps[]
): NonNullable<HelmetProps[T]> => {
  // Calculate list of tags, giving priority innermost component (end of the propslist)
  const approvedSeenTags = {};

  return propsList
    .filter(props => {
      if (Array.isArray(props[tagName])) {
        return true;
      }
      if (typeof props[tagName] !== 'undefined') {
        warn(
          `Helmet: ${tagName} should be of type "Array". Instead found type "${typeof props[
            tagName
          ]}"`
        );
      }
      return false;
    })
    .map(props => props[tagName])
    .reverse()
    .reduce<HelmetProps[T][]>((approvedTags, instanceTags) => {
      const instanceSeenTags: Record<string, Record<string, boolean>> = {};

      instanceTags!
        .filter(tag => {
          type TagAttributeKey = keyof UnionToIntersection<typeof tag>;

          let primaryAttributeKey: TagAttributeKey | undefined;

          for (const attributeKey of Object.keys(tag) as TagAttributeKey[]) {
            const lowerCaseAttributeKey = attributeKey; // .toLowerCase() as TagAttributeKey;

            // Special rule with link tags, since rel and href are both primary tags, rel takes priority
            if (
              primaryAttributes.includes(lowerCaseAttributeKey) &&
              !(
                primaryAttributeKey === TAG_PROPERTIES.REL &&
                tag[primaryAttributeKey].toLowerCase() === 'canonical'
              ) &&
              !(
                lowerCaseAttributeKey === TAG_PROPERTIES.REL &&
                tag[lowerCaseAttributeKey].toLowerCase() === 'stylesheet'
              )
            ) {
              primaryAttributeKey = lowerCaseAttributeKey;
            }
            // Special case for innerHTML which doesn't work lowercased
            if (
              primaryAttributes.includes(attributeKey) &&
              (attributeKey === TAG_PROPERTIES.INNER_HTML ||
                attributeKey === TAG_PROPERTIES.CSS_TEXT ||
                attributeKey === TAG_PROPERTIES.ITEM_PROP)
            ) {
              primaryAttributeKey = attributeKey as TagAttributeKey;
            }
          }

          if (!primaryAttributeKey || !tag[primaryAttributeKey]) {
            // If we don't have a primary attribute for this tag, or that tag doesn't have a
            // value for the primary attribute then it's an invalid tag and we don't use it.
            return false;
          }

          const value = tag[primaryAttributeKey].toLowerCase();

          if (!approvedSeenTags[primaryAttributeKey]) {
            approvedSeenTags[primaryAttributeKey] = {};
          }

          if (!instanceSeenTags[primaryAttributeKey]) {
            instanceSeenTags[primaryAttributeKey] = {};
          }

          if (!approvedSeenTags[primaryAttributeKey][value]) {
            instanceSeenTags[primaryAttributeKey][value] = true;
            return true;
          }

          return false;
        })
        .reverse()
        .forEach(tag => approvedTags.push(tag));

      // Update seen tags with tags from this instance
      for (const attributeKey of Object.keys(instanceSeenTags)) {
        approvedSeenTags[attributeKey] = {
          ...approvedSeenTags[attributeKey],
          ...instanceSeenTags[attributeKey],
        };
      }

      return approvedTags;
    }, [])
    .reverse();
};

const getAnyTruthyFromPropsList = (
  propsList: HelmetProps[],
  checkedTag: keyof HelmetProps
): boolean => {
  return propsList.some(props => props[checkedTag]);
};

export interface HelmetInternalState {
  baseTag: React.JSX.IntrinsicElements['base'][];
  bodyAttributes: NonNullable<BodyProps>;
  defer?: boolean;
  encode?: boolean;
  htmlAttributes: NonNullable<HtmlProps>;
  linkTags: React.JSX.IntrinsicElements['link'][];
  metaTags: React.JSX.IntrinsicElements['meta'][];
  noscriptTags: { innerHTML: string }[];
  onChangeClientState?: NonNullable<HelmetProps['onChangeClientState']>;
  scriptTags: React.JSX.IntrinsicElements['script'][];
  styleTags: { cssText: string }[];
  title: string | undefined;
  titleAttributes: { itemprop?: string | undefined };
  prioritizeSeoTags?: boolean;
}

export const reducePropsToState = (propsList: HelmetProps[]): HelmetInternalState => {
  return {
    baseTag: getBaseTagFromPropsList(propsList),
    bodyAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.BODY, propsList),
    defer: getInnermostProperty(propsList, HELMET_PROPS.DEFER),
    encode: getInnermostProperty(propsList, HELMET_PROPS.ENCODE_SPECIAL_CHARACTERS),
    htmlAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.HTML, propsList),
    linkTags: getTagsFromPropsList(
      TAG_NAMES.LINK,
      [TAG_PROPERTIES.REL, TAG_PROPERTIES.HREF],
      propsList
    ),
    metaTags: getTagsFromPropsList(
      TAG_NAMES.META,
      [
        TAG_PROPERTIES.NAME,
        TAG_PROPERTIES.CHARSET,
        TAG_PROPERTIES.HTTPEQUIV,
        TAG_PROPERTIES.PROPERTY,
        TAG_PROPERTIES.ITEM_PROP,
      ],
      propsList
    ),
    noscriptTags: getTagsFromPropsList(TAG_NAMES.NOSCRIPT, [TAG_PROPERTIES.INNER_HTML], propsList),
    onChangeClientState: getOnChangeClientState(propsList),
    scriptTags: getTagsFromPropsList(
      TAG_NAMES.SCRIPT,
      [TAG_PROPERTIES.SRC, TAG_PROPERTIES.INNER_HTML],
      propsList
    ),
    styleTags: getTagsFromPropsList(TAG_NAMES.STYLE, [TAG_PROPERTIES.CSS_TEXT], propsList),
    title: getTitleFromPropsList(propsList),
    titleAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.TITLE, propsList),
    prioritizeSeoTags: getAnyTruthyFromPropsList(propsList, HELMET_PROPS.PRIORITIZE_SEO_TAGS),
  };
};

export function flattenArray(possibleArray: string): string;
export function flattenArray(possibleArray: string | undefined): string | undefined;
export function flattenArray(possibleArray: string[]): string;
export function flattenArray(possibleArray: string | string[] | undefined): string | undefined {
  return Array.isArray(possibleArray) ? possibleArray.join('') : possibleArray;
}

type SeoPriorityOptions = 'metaTags' | 'linkTags' | 'scriptTags';

const checkIfPropsMatch = (props: HelmetInternalState[SeoPriorityOptions][number], toMatch) => {
  const keys = Object.keys(props);
  for (let i = 0; i < keys.length; i += 1) {
    // e.g. if rel exists in the list of allowed props [amphtml, alternate, etc]
    if (toMatch[keys[i]] && toMatch[keys[i]].includes(props[keys[i]])) {
      return true;
    }
  }
  return false;
};

export const prioritizer = <T extends HelmetInternalState[SeoPriorityOptions][number]>(
  elementsList: T[],
  propsToMatch
) => {
  return elementsList.reduce<{ priority: T[]; default: T[] }>(
    (acc, elementAttrs) => {
      if (checkIfPropsMatch(elementAttrs, propsToMatch)) {
        acc.priority.push(elementAttrs);
      } else {
        acc.default.push(elementAttrs);
      }
      return acc;
    },
    { priority: [], default: [] }
  );
};

export const without = (obj, key) => {
  return {
    ...obj,
    [key]: undefined,
  };
};
