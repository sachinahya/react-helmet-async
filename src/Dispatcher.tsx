import { Component } from 'react';
import shallowEqual from 'shallowequal';
import handleStateChangeOnClient from './client';
import mapStateOnServer from './server';
import { reducePropsToState } from './utils';
import Provider, { HelmetContext } from './Provider';
import { HelmetProps } from './index';

export interface DispatcherProps extends HelmetProps {
  context: HelmetContext;
}

export default class Dispatcher extends Component<DispatcherProps> {
  rendered = false;

  shouldComponentUpdate(nextProps) {
    return !shallowEqual(nextProps, this.props);
  }

  componentDidUpdate() {
    this.emitChange();
  }

  componentWillUnmount() {
    const { helmetInstances } = this.props.context;
    helmetInstances.remove(this);
    this.emitChange();
  }

  emitChange() {
    const { helmetInstances, setHelmet } = this.props.context;
    let serverState = null;
    const propsList = helmetInstances.get().map(instance => {
      const props = { ...instance.props };
      delete props.context;
      return props;
    });
    const state = reducePropsToState(propsList);
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      serverState = mapStateOnServer(state);
    }
    setHelmet(serverState);
  }

  // componentWillMount will be deprecated
  // for SSR, initialize on first render
  // constructor is also unsafe in StrictMode
  init() {
    if (this.rendered) {
      return;
    }

    this.rendered = true;

    const { helmetInstances } = this.props.context;
    helmetInstances.add(this);
    this.emitChange();
  }

  render() {
    this.init();

    return null;
  }
}
