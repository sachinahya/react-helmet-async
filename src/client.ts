import { HELMET_ATTRIBUTE, TAG_NAMES, TAG_PROPERTIES } from './constants';
import { ReducedState, flattenArray } from './utils';

const updateTags = <
  T extends Omit<typeof TAG_NAMES, 'FRAGMENT'>[keyof Omit<typeof TAG_NAMES, 'FRAGMENT'>],
>(
  type: T,
  tags: ReducedState[
    | 'baseTag'
    | 'linkTags'
    | 'metaTags'
    | 'noscriptTags'
    | 'scriptTags'
    | 'styleTags']
) => {
  type Tag = (typeof tags)[number];

  const headElement = document.head || document.querySelector(TAG_NAMES.HEAD);
  const tagNodes = headElement.querySelectorAll<HTMLElementTagNameMap[T]>(
    `${type}[${HELMET_ATTRIBUTE}]`
  );
  const oldTags: HTMLElementTagNameMap[T][] = [].slice.call(tagNodes);
  const newTags: HTMLElementTagNameMap[T][] = [];
  let indexToDelete: number;

  if (tags && tags.length) {
    tags.forEach((tag: Tag) => {
      const newElement = document.createElement<T>(type);

      // eslint-disable-next-line
      for (const attribute in tag) {
        if (Object.prototype.hasOwnProperty.call(tag, attribute)) {
          if (attribute === TAG_PROPERTIES.INNER_HTML) {
            newElement.innerHTML = tag.innerHTML;
          } else if (attribute === TAG_PROPERTIES.CSS_TEXT) {
            if (newElement.styleSheet) {
              newElement.styleSheet.cssText = tag.cssText;
            } else {
              newElement.appendChild(document.createTextNode(tag.cssText));
            }
          } else {
            const value = typeof tag[attribute] === 'undefined' ? '' : tag[attribute];
            newElement.setAttribute(attribute, value);
          }
        }
      }

      newElement.setAttribute(HELMET_ATTRIBUTE, 'true');

      // Remove a duplicate tag from domTagstoRemove, so it isn't cleared.
      if (
        oldTags.some((existingTag, index) => {
          indexToDelete = index;
          return newElement.isEqualNode(existingTag);
        })
      ) {
        oldTags.splice(indexToDelete, 1);
      } else {
        newTags.push(newElement);
      }
    });
  }

  oldTags.forEach(tag => tag.parentNode.removeChild(tag));
  newTags.forEach(tag => headElement.appendChild(tag));

  return {
    oldTags,
    newTags,
  };
};

const updateAttributes = <T extends keyof ReducedState>(
  tagName: (typeof TAG_NAMES)[keyof typeof TAG_NAMES],
  attributes: ReducedState[T]
): void => {
  const elementTag = document.getElementsByTagName(tagName)[0];

  if (!elementTag) {
    return;
  }

  const helmetAttributeString = elementTag.getAttribute(HELMET_ATTRIBUTE);
  const helmetAttributes = helmetAttributeString ? helmetAttributeString.split(',') : [];
  const attributesToRemove = [].concat(helmetAttributes);
  const attributeKeys = Object.keys(attributes);

  for (let i = 0; i < attributeKeys.length; i += 1) {
    const attribute = attributeKeys[i];
    const value = attributes[attribute] || '';

    if (elementTag.getAttribute(attribute) !== value) {
      elementTag.setAttribute(attribute, value);
    }

    if (helmetAttributes.indexOf(attribute) === -1) {
      helmetAttributes.push(attribute);
    }

    const indexToSave = attributesToRemove.indexOf(attribute);
    if (indexToSave !== -1) {
      attributesToRemove.splice(indexToSave, 1);
    }
  }

  for (let i = attributesToRemove.length - 1; i >= 0; i -= 1) {
    elementTag.removeAttribute(attributesToRemove[i]);
  }

  if (helmetAttributes.length === attributesToRemove.length) {
    elementTag.removeAttribute(HELMET_ATTRIBUTE);
  } else if (elementTag.getAttribute(HELMET_ATTRIBUTE) !== attributeKeys.join(',')) {
    elementTag.setAttribute(HELMET_ATTRIBUTE, attributeKeys.join(','));
  }
};

const updateTitle = (title: string, attributes: ReducedState['titleAttributes']) => {
  if (typeof title !== 'undefined' && document.title !== title) {
    document.title = flattenArray(title);
  }

  updateAttributes(TAG_NAMES.TITLE, attributes);
};

const commitTagChanges = (newState: ReducedState, cb?: () => void) => {
  const {
    baseTag,
    bodyAttributes,
    htmlAttributes,
    linkTags,
    metaTags,
    noscriptTags,
    onChangeClientState,
    scriptTags,
    styleTags,
    title,
    titleAttributes,
  } = newState;
  updateAttributes(TAG_NAMES.BODY, bodyAttributes);
  updateAttributes(TAG_NAMES.HTML, htmlAttributes);

  updateTitle(title, titleAttributes);

  const tagUpdates = {
    baseTag: updateTags(TAG_NAMES.BASE, baseTag),
    linkTags: updateTags(TAG_NAMES.LINK, linkTags),
    metaTags: updateTags(TAG_NAMES.META, metaTags),
    noscriptTags: updateTags(TAG_NAMES.NOSCRIPT, noscriptTags),
    scriptTags: updateTags(TAG_NAMES.SCRIPT, scriptTags),
    styleTags: updateTags(TAG_NAMES.STYLE, styleTags),
  };

  const addedTags = {};
  const removedTags = {};

  Object.keys(tagUpdates).forEach(tagType => {
    const { newTags, oldTags } = tagUpdates[tagType];

    if (newTags.length) {
      addedTags[tagType] = newTags;
    }
    if (oldTags.length) {
      removedTags[tagType] = tagUpdates[tagType].oldTags;
    }
  });

  if (cb) {
    cb();
  }

  onChangeClientState(newState, addedTags, removedTags);
};

// eslint-disable-next-line
let _helmetCallback = null;

const handleStateChangeOnClient = (newState: ReducedState) => {
  if (_helmetCallback) {
    cancelAnimationFrame(_helmetCallback);
  }

  if (newState.defer) {
    _helmetCallback = requestAnimationFrame(() => {
      commitTagChanges(newState, () => {
        _helmetCallback = null;
      });
    });
  } else {
    commitTagChanges(newState);
    _helmetCallback = null;
  }
};

export default handleStateChangeOnClient;
