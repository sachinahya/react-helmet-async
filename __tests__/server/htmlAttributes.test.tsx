import React from 'react';
import ReactServer from 'react-dom/server';
import { Helmet } from '../../src';
import Provider from '../../src/Provider';
import { render } from './utils';

Helmet.defaultProps.defer = false;

beforeAll(() => {
  Provider.canUseDOM = false;
});

afterAll(() => {
  Provider.canUseDOM = true;
});

describe('server', () => {
  describe('API', () => {
    it('renders html attributes as component', () => {
      const context = {};
      render(
        <Helmet
          htmlAttributes={{
            lang: 'ga',
            className: 'myClassName',
          }}
        />,
        context
      );

      const { htmlAttributes } = context.helmet;
      const attrs = htmlAttributes.toComponent();

      expect(attrs).toBeDefined();

      const markup = ReactServer.renderToStaticMarkup(<html lang="en" {...attrs} />);

      expect(markup).toMatchSnapshot();
    });

    it('renders html attributes as string', () => {
      const context = {};
      render(
        <Helmet
          htmlAttributes={{
            lang: 'ga',
            class: 'myClassName',
          }}
        />,
        context
      );

      const head = context.helmet;

      expect(head.htmlAttributes).toBeDefined();
      expect(head.htmlAttributes.toString).toBeDefined();
      expect(head.htmlAttributes.toString()).toMatchSnapshot();
    });
  });

  describe('Declarative API', () => {
    it('renders html attributes as component', () => {
      const context = {};
      render(
        <Helmet>
          <html lang="ga" className="myClassName" />
        </Helmet>,
        context
      );

      const { htmlAttributes } = context.helmet;
      const attrs = htmlAttributes.toComponent();

      expect(attrs).toBeDefined();

      const markup = ReactServer.renderToStaticMarkup(<html lang="en" {...attrs} />);

      expect(markup).toMatchSnapshot();
    });

    it('renders html attributes as string', () => {
      const context = {};
      render(
        <Helmet>
          <html lang="ga" className="myClassName" />
        </Helmet>,
        context
      );

      const head = context.helmet;

      expect(head.htmlAttributes).toBeDefined();
      expect(head.htmlAttributes.toString).toBeDefined();
      expect(head.htmlAttributes.toString()).toMatchSnapshot();
    });
  });
});
