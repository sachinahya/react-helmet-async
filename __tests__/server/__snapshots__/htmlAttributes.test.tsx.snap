// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server API renders html attributes as component 1`] = `"<html lang=\\"ga\\" class=\\"myClassName\\"></html>"`;

exports[`server API renders html attributes as string 1`] = `"lang=\\"ga\\" class=\\"myClassName\\""`;

exports[`server Declarative API renders html attributes as component 1`] = `"<html lang=\\"ga\\" class=\\"myClassName\\"></html>"`;

exports[`server Declarative API renders html attributes as string 1`] = `"lang=\\"ga\\" class=\\"myClassName\\""`;
